### PR TITLE
Add repo secret example using libsodium

### LINKED ISSUE #1607: Add an integration test or example on how to use the addOrUpdateRepoSecret function.
I wanted to use this library to push secrets from my env to my github repo easily, and eventually got it figured out. But It would be very helpful to have an example of how to do that since it requires sodiumlib which is extremely difficult to parse if you are trying to figure out the bare minimum needed to encrypt a secret for posting to github.

essentially what I had to do to update the repo's secret was

Authenticate with github

```go
func GithubAuth() (context.Context, *github.Client, error) {
	token := os.Getenv("API_GITHUB_TOKEN")
	if token == "" {
		return nil, nil, errors.New(`no API_GITHUB_TOKNE was found in the environment variables.
This is needed to authenticate with github. Please generate a github token that has access to do what you are trying to do
and export it as an environment variable.

ex:
	export API_GITHUB_TOKEN="<token contents string>"
`)
	}
	ctx := context.Background()
	ts := oauth2.StaticTokenSource(
		&oauth2.Token{AccessToken: token},
	)
	tc := oauth2.NewClient(ctx, ts)

	client := github.NewClient(tc)
	return ctx, client, nil
}
```

Get the base64 encoded public key of the repo to encrypt a secret using sodiumlib

```go
publicKey, _, err := client.Actions.GetRepoPublicKey(ctx, owner, repo)
```

encrypt the secret string with the public key

```go
func encryptSecretWithPublicKey(publicKey *github.PublicKey, secretName string, secretValue string) (*github.EncryptedSecret, error) {
	decodedPublicKey, err := base64.StdEncoding.DecodeString(publicKey.GetKey())
	if err != nil {
		return nil, errors.New(fmt.Sprintf("base64.StdEncoding.DecodeString was unable to decode public key: %v", err))
	}

	secretBytes := []byte(secretValue)
	encryptedBytes, exit := sodium.CryptoBoxSeal(secretBytes, decodedPublicKey)
	if exit != 0 {
		return nil, errors.New("sodium.CryptoBoxSeal exited with non zero exit code")
	}

	encryptedString := base64.StdEncoding.EncodeToString(encryptedBytes)
	keyID := publicKey.GetKeyID()
	encryptedSecret := &github.EncryptedSecret{
		Name:           secretName,
		KeyID:          keyID,
		EncryptedValue: encryptedString,
	}
	return encryptedSecret, nil
}
```

bring it all together with AddRepoSecret function

```go
func AddRepoSecret(owner string, repo string, secretName string, secretValue string) (string, error) {
	ctx, client, err := GithubAuth()
	if err != nil {
		return "", err
	}
	publicKey, _, err := client.Actions.GetRepoPublicKey(ctx, owner, repo)
	if err != nil {
		return "", err
	}

	encryptedSecret, err := encryptSecretWithPublicKey(publicKey, secretName, secretValue)
	if err != nil {
		return "", err
	}

	_, err = client.Actions.CreateOrUpdateRepoSecret(ctx, owner, repo, encryptedSecret)
	if err != nil {
		return "", errors.New(fmt.Sprintf("Actions.CreateOrUpdateRepoSecret returned error: %v", err))
	}

	return secretName, nil
}
```

Hopefully this helps someone, if there is a direction I should take I can work to contribute this back in some form or another.

### PR DESCRIPTION
**Pull request by raboley**
Added a new Example using libsodium to create a new secret in a github repo for use in github actions.

It is usable as a cli package to produce a secret in a given repo that the user owns.

usage is
main creates a new secret in github for a given owner/repo based on a secretName passed in as an arg
and a github token, and secret value provided via the environment

usage:
	export GITHUB_AUTH_TOKEN=<auth token from github that has secret create rights>
	export SECRET_VARIABLE=<secret value of the secret variable>
	go run main.go -owner <owner name> -repo <repository name> SECRET_VARIABLE

ex:
	export GITHUB_AUTH_TOKEN=0000000000000000
	export SECRET_VARIABLE="my-secret"
	go run main.go -owner google -repo go-github SECRET_VARIABLE

Once it runs go to the github repository > settings > left side options bar > Secrets
And you should see the new secret appear there.

### CODE REVIEW

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:1]: To follow idiomatic Go style and to work nicely with `godoc`, please put the package comment here. For example:

```
// newreposecret creates a new secret in GitHub for a given ...
// 
// Usage:
// ...
//
// Example:
// ...
package main
```

Please write this with the Godocs in mind... use complete sentences ending in periods. Capitalize new sentences, capitalize GitHub correctly, etc.

Then you can completely delete lines 21-37 below.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:43]: Let's please make each `log.Fatal` message just one line long if possible. Hopefully the documentation up at the top or the comments will explain all the details

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:74]:
```suggestion
		log.Fatal(err)
```

here and elsewhere

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:70]:
```suggestion
		log.Fatalf("unable to authorize using env GITHUB_AUTH_TOKEN: %v", err)
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:87]:
```suggestion
	fmt.Printf("Added secret %q to the repo %v/%v\n", addedSecretName, *owner, *repo)
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:104]: Again, the error is way to long.  Please make it short and sweet, and use `fmt.Errorf`.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:98]: These last 3 lines are unnecessary, and since you are using the `flag` package, please don't use `os` as well... just use `flag.NArgs` and `flag.Arg` here.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:117]: This method does not need exporting.  But if it did, it would be spelled `GitHubAuth`.  Since it isn't, it should be `githubAuth`.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:120]: Another error message that is way too long. Short and sweet, please.
Don't make your error messages tutorials... the documentation should explain the usage, the error message should just point out the problem.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:138]: This comment is super-hard to read.  Please break it up into paragraphs with a blank comment line separating the paragraphs.
Also, please capitalize `GitHub` properly and make complete sentences ending in periods.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:158]: This method does not need exporting, so it will be `addRepoSecret`.  Also, there is no reason to return the passed-in secretName, so don't.

```suggestion
func addRepoSecret(ctx context.Context, client *github.Client, owner, repo, secretName, secretValue string) error {
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:180]: `fmt.Errorf`

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:177]:
```suggestion
func encryptSecretWithPublicKey(publicKey *github.PublicKey, secretName, secretValue string) (*github.EncryptedSecret, error) {
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @raboley.  This is a good start, but let's please clean it up a bit. This may take a few iterations.

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:177]: Is this intended to remove the typing of the secret name parameter? that is the only change I can see here and I just wanted to make sure that is what you intended. I don't off top of my head understand why we would want to do that, but am open to reasons why!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:22]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated! point taken that makes sense!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: yeah this was a nightmare after re-reading it. I rewrote it and spaced things out more which is hopefully more readable. It is all under the hood stuff too, so I mostly intended it to explain the process so that if people look at it they can hopefully understand why it has to be so complicated.

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: unexported and removed the return. good idea!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:162]: Learning about fmt.Errorf has been worth it in itself by submitting this PR. Thank you for showing me it!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:22]: I went through and updated all the places you mentioned to update and pushed a new commit. Thank you so much for taking the time to review my code so thoroughly and precisely. I really appreciate you taking the time to clean this up with me. I'm no pro gopher so I really appreciate everything you are doing to work with me! I wasn't sure if I should resolve the conversations or you, so I left them open so that you could resolve them if my updates met your expectations. Thank you again for all your work on this! I'm really excited to potentially get this merged some time 😃

**Comment by raboley** [comment]
> Thank you, @raboley. This is a good start, but let's please clean it up a bit. This may take a few iterations.

This is all really great @gmlewis thank you so much for the feedback I have updated based on your comments!

File: example/newreposecret/main.go
Comment (gmlewis) [reply example/newreposecret/main.go:159]: Ah, sorry for dropping this suggestion without explaining it...

In Go, if two or more successive arguments to a method all have the same type, you can combine them together (which is the recommended idiomatic Go style), so:

```
func a(b int, c int, d int, e int, h string, i string j string) {}
```

becomes:

```
func a(b, c, d, e int, h, i, j string) {}
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:1]:
```suggestion
// Copyright 2020 The go-github AUTHORS. All rights reserved.
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:6]:
```suggestion
// newreposecret creates a new secret in GitHub for a given owner/repo.
//
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:9]:
```suggestion
// To authenticate with GitHub provide it via an environment variable GITHUB_AUTH_TOKEN.
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:11]:
```suggestion
// To verify the new secret, navigate to GitHub Repository > Settings > left side options bar > Secrets.
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:30]: Did you run this code through `gofmt`?  I'm surprised the GitHub workflow tests did not catch this!  In fact, I'm going to stop this code review here and give you my feedback so far and see if I can fix this.

Just FYI... I use Emacs primarily, vim sometimes, and VSCode every now and then... all are great editors, and the last one has amazing Go support.  Whatever editor(s) you use, you should have them set to "format on save" or run "goimports" to update your imports and also fix the formatting... it's a huge time-saver!

**Review by gmlewis — CHANGES_REQUESTED** [review]

File: example/newreposecret/main.go
Comment (gmlewis) [reply example/newreposecret/main.go:30]: OK, Today I Learned that "goimports" is the one that nicely splits up imports into standard library and external ones with a blank line in between.  I thought that was "gofmt", but I was wrong.  If you could please add a blank line between standard library imports and external imports, or just use "goimports", that would be great, so that our repo is consistent... thank you!

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:55]:
```suggestion
		log.Fatal("please provide required flag --owner to specify GitHub user/org owner")
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:73]:
```suggestion
	if err := addRepoSecret(ctx, client, *owner, *repo, secretName, secretValue); err != nil {
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:74]: this line can be removed (after adding it to the previous line).

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:84]:
```suggestion
		return "", fmt.Errorf("missing argument secret name")
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:85]: This line can be deleted.

**Review by gmlewis — CHANGES_REQUESTED** [review]
more later...

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:116]:
```suggestion
// To communicate with GitHub, addRepoSecret requires a client and context.
```

although I think this can be deleted completely as it is rather obvious.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:117]:
```suggestion
// To determine what repository to add this secret to, it will use the owner and repo combination.
```

This is obvious too, and I think it can be deleted.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:118]:
```suggestion
// Finally the secretName and secretValue will determine the name of the secret added and it's corresponding value
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
I'm interrupted again, but I'm noticing that your sentences all need to be fully formed ending with periods for the resulting auto-generated go-docs.

I'm going to hold off on further reviewing this PR until some more updates are made so that gives you a chance to catch up.  Thanks.

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:159]: Ahh thank you so much for the explanation! I thought it was a mistake or something, but now I learned something new! Updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:33]: I had never heard of goimports, so I added that to my idea (Jetbrains Goland) so it should catch this in the future!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go:74]: Is this how I should do all my errs? or is this just because err is defined above?

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: updated!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: deleted!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: deleted!

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: I can't see the difference here, but I added a period at the end since it is a full sentence.

**Comment by raboley** [comment]
I updated based on all the latest comments, but I will take another look over all the prose to ensure it is punctuated, spelled and cased correctly.

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:116]: Let's please change this URL so it is not just for Mac:

```suggestion
// https://github.com/jedisct1/libsodium
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:144]: I'm going to recommend this again to keep the code a bit more concise.
You typically do this when there is not another value that you are keeping around for other use.

```suggestion
	if _, err := client.Actions.CreateOrUpdateRepoSecret(ctx, owner, repo, encryptedSecret); err != nil {
```

File: example/newreposecret/main.go
Comment (gmlewis) [inline example/newreposecret/main.go:145]: The this line can be deleted.

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @raboley ... this is looking great!
Just a couple more minor tweaks, please... then we should be ready for a second review and merging.

**Comment by raboley** [comment]
> Thank you, @raboley ... this is looking great!
> Just a couple more minor tweaks, please... then we should be ready for a second review and merging.

Thank you again so much for the thoughtful edits @gmlewis! I had my wife (a teacher and much better writer than I) help me rewrite some of my comments to be full sentences that are more clear. I don't believe I have any sentence fragments or missing punctuation! I also made the other update around simplifying the if statement. 

commit hash of the comment rewrites: 1bc4e4c8676108c8608971a8e282bedd8a25d898

**Review by gmlewis — APPROVED** [review]
Apart from the one comment, this now LGTM.
Please remove the redundant sentence, then we will be ready for a second LGTM and merging.

Thank you, @raboley !

File: example/newreposecret/main.go
Comment (raboley) [reply example/newreposecret/main.go]: Amazing! Thank you so much for all your feedback and assistance! I learned a ton and will strive to be a better gopher going forward.

Comment (raboley) [reply example/newreposecret/main.go]: @gmlewis I removed the duplicative lines!

**Comment by raboley** [comment]
@wesleimp Pleased to meet you! Let me know of any improvements I can make and I will try to do so as speedily as possible!

**Review by gmlewis — APPROVED** [review]

**Comment by raboley** [comment]
@wesleimp let me know if there is anything I can do to help in your code review! Thank you again for taking the time and I hope everything is going well with you.

**Review by wesleimp — APPROVED** [review]
Thank you, @raboley 
LGTM 👌🏼 

Sorry for my late, my time is so limited on these days. :grimmacing:

**Comment by raboley** [comment]
Thank you!! 😄

**Comment by raboley** [comment]
Totally understand @wesleimp I really appreciate you taking the time! I know it is in short supply for us all! Is there anything else I need to do prior to this being merged or are we at the end?

**Comment by wesleimp** [comment]
@raboley we are at the end. Unfortunately I don't have permission to merge PRs.

Ping @gmlewis :)

**Comment by gmlewis** [comment]
Successfully pinged.  😂 
Thank you, @raboley and @wesleimp !
Merging.

**Pull request merged**
